confluent_kafka/avro/serializer/message_serializer.py

```diff
@@ -101,34 +108,35 @@ def encode_record_with_schema(self, topic, schema, record, is_key=False):
             raise serialize_err(message)
 
         # cache writer
-        self.id_to_writers[schema_id] = avro.io.DatumWriter(schema)
+        if id not in self.id_to_writers:
+            self.id_to_writers[schema_id] = self.make_datum_writer(schema)
 
         return self.encode_record_with_schema_id(schema_id, record, is_key=is_key)
 
-    def encode_record_with_schema_id(self, schema_id, record, is_key=False):
-        """
-        Encode a record with a given schema id.  The record must
-        be a python dictionary.
-        @:param: schema_id : integer ID
-        @:param: record : An object to serialize
-        @:param is_key : If the record is a key
-        @:returns: decoder function
-        """
+    def _ensure_writer(self, schema_id, is_key=False):
         serialize_err = KeySerializerError if is_key else ValueSerializerError
 
-        # use slow avro
         if schema_id not in self.id_to_writers:
             # get the writer + schema
-
             try:
                 schema = self.registry_client.get_by_id(schema_id)
                 if not schema:
                     raise serialize_err("Schema does not exist")
-                self.id_to_writers[schema_id] = avro.io.DatumWriter(schema)
+                self.id_to_writers[schema_id] = self.make_datum_writer(schema)
```

Comment:
I would move this logic back into `_encode_record_with_schema_id` and eliminate the redundant `_slow_encode_record_with_schema_id` and `_fast_encode_record_with_schema_id` methods. If we provide the same interface for both writers there is no longer a need to maintain two code paths which is a bit more DRY.